### PR TITLE
refactor next slot cache

### PR DESCRIPTION
**Pull request by potuz**
This PR refactors the next slot cache to keep track of two alternative states. This makes the cache safe under skipped slots and 1-slot reorgs, including across epoch boundaries. Design is in https://www.notion.so/arbitrum/Next-Slot-Cache-2-0-301e3788c1b248e6a587bd3505c4034a

This also fixes some bugs making the cache not thread safe  (by copying the root slice instead of keeping the slice parameter in the cache).

### CODE REVIEW

File: beacon-chain/core/transition/transition.go [hunk 1]
```diff
@@ -147,25 +147,15 @@ func ProcessSlotsUsingNextSlotCache(
 	ctx, span := trace.StartSpan(ctx, "core.state.ProcessSlotsUsingNextSlotCache")
 	defer span.End()
 
-	// Check whether the parent state has been advanced by 1 slot in next slot cache.
```
Comment (potuz) [inline beacon-chain/core/transition/transition.go:150]: This is a very simple function that was overly commented making it less readable. I removed some useless variables and most comments.

File: beacon-chain/blockchain/process_block.go [hunk 2]
```diff
@@ -697,21 +697,31 @@ func (s *Service) fillMissingBlockPayloadId(ctx context.Context) error {
...
 	headBlock, err := s.headBlock()
 	if err != nil {
 		s.headLock.RUnlock()
-		return err
+		log.WithError(err).Debug("could not perform late block tasks: failed to retrieve head block")
+		return
 	}
-	headState := s.headState(ctx)
 	headRoot := s.headRoot()
+	headState := s.headState(ctx)
```
Comment (nisdas) [inline beacon-chain/blockchain/process_block.go:710]: Is there a reason you switched the ordering from 
- `state then root` to `root and then state` ?

**Review by terencechain — APPROVED** [review]

**Review by nisdas — COMMENTED** [review]
How does this prevent updates to the cache for non-canonical blocks ? `UpdateNextSlotCache` is still called for all blocks processed

**Comment by potuz** [comment]
> How does this prevent updates to the cache for non-canonical blocks ? `UpdateNextSlotCache` is still called for all blocks processed

I could prevent it, but decided that it's more useful to actually keep them, we may get reorgs that are non-canonical as soon as we insert them if we receive them near the 4 second boundary and they become canonical at 12 seconds. The likelyhood of this happening is higher than the likelihood of the previous state being needed advanced to another slot to pass an epoch transition, so I decided to keep them here.

**Review by nisdas — APPROVED** [review]

**Pull request merged**